### PR TITLE
handle user printed json output in stdout

### PR DESCRIPTION
**Pull request by devoncarew**
- fix https://github.com/flutter/flutter-intellij/issues/1879, fix https://github.com/flutter/flutter-intellij/issues/1615
- harden the daemon protocol against user printed `[{...}]` json blocks

### CODE REVIEW

File: src/io/flutter/run/daemon/DaemonApi.java [hunk 1]
```diff
@@ -236,6 +203,64 @@ private Command takePending(int id) {
     }
   }
 
+  /**
+   * Parse the given string; if it's valid JSON, and a valid Daemon response of event, then return
```
Comment (pq) [inline src/io/flutter/run/daemon/DaemonApi.java]: Not important probably but I find this description hard to follow... 

> "valid Daemon response of event" 

=> missing a word maybe?

Comment (pq) [reply src/io/flutter/run/daemon/DaemonApi.java]: Or an extra one?

File: src/io/flutter/run/daemon/DaemonApi.java [hunk 2]
```diff
@@ -236,6 +203,64 @@ private Command takePending(int id) {
...
+   * the parsed JsonObject.
+   */
+  static JsonObject parseAndValidateDaemonEvent(String str) {
```
Comment (pq) [inline src/io/flutter/run/daemon/DaemonApi.java]: Maybe renaming `str` to `event` would help here?

File: src/io/flutter/run/daemon/DaemonApi.java [hunk 3]
```diff
@@ -236,6 +203,64 @@ private Command takePending(int id) {
...
+    if (!str.startsWith("[{")) {
+      return null;
+    }
+
+    str = str.trim();
```
Comment (pq) [inline src/io/flutter/run/daemon/DaemonApi.java]: Should the trim come first?

File: src/io/flutter/run/daemon/DaemonApi.java (hunk 2 above)
Comment (pq) [reply src/io/flutter/run/daemon/DaemonApi.java]: Or `eventString`...

File: src/io/flutter/run/daemon/DaemonConsoleView.java [hunk 4]
```diff
@@ -82,15 +82,15 @@ public void print(@NotNull String text, @NotNull ConsoleViewContentType contentT
...
+      if (DaemonApi.parseAndValidateDaemonEvent(line.trim()) != null) {
+        if (FlutterSettings.getInstance().isVerboseLogging()) {
+          LOG.info(line.trim().trim());
+        }
         return;
       }
       else {
-        // We're seeing a spurious newline before some launches; this removed any single
-        // newline that occur before we've printed text.
+        // We're seeing a spurious newline before some launches; this removed any single newline that occur
```
Comment (pq) [inline src/io/flutter/run/daemon/DaemonConsoleView.java]: Also not important but something doesn't agree with me.

Maybe:

this removes any single newline that occurs

or 

this removes any single newlines that occur

or

this removes any single newline that occured

File: src/io/flutter/run/daemon/DaemonEvent.java [hunk 5]
```diff
@@ -227,7 +240,8 @@ void accept(Listener listener) {
 
     private Boolean finished;
 
-    @NotNull String getType() {
+    @NotNull
+    String getType() {
       return progressId == null ? "" : progressId;
```
Comment (pq) [inline src/io/flutter/run/daemon/DaemonEvent.java]: FWIW: `StringUtil.notNullize(progressId);` is another cute/idiomatic way to do this.

File: src/io/flutter/run/daemon/DaemonEvent.java [hunk 6]
```diff
@@ -21,26 +21,25 @@
...
 abstract class DaemonEvent {
-
   /**
    * Parses an event and sends it to the listener.
    */
   static void dispatch(@NotNull JsonObject obj, @NotNull Listener listener) {
     final JsonPrimitive primEvent = obj.getAsJsonPrimitive("event");
     if (primEvent == null) {
-      LOG.error("Missing event field in JSON from flutter process: " + obj);
+      LOG.info("Missing event field in JSON from flutter process: " + obj);
```
Comment (jacob314) [inline src/io/flutter/run/daemon/DaemonEvent.java]: why are these going from error to info log messages?

File: testSrc/unit/io/flutter/run/daemon/DaemonApiTest.java [hunk 7]
```diff
@@ -74,7 +75,7 @@ public void canEnableDeviceEvents() throws Exception {
     checkLog("{\"method\":\"device.enable\",\"id\":0}");
```
Comment (jacob314) [inline testSrc/unit/io/flutter/run/daemon/DaemonApiTest.java]: an you add a test case matching the new edge cases that are now handled correctly?

File: src/io/flutter/run/daemon/DaemonConsoleView.java [hunk 8]
```diff
@@ -82,15 +82,15 @@ public void print(@NotNull String text, @NotNull ConsoleViewContentType contentT
 
   private void writeAvailableLines() {
     for (String line : stdoutParser.getAvailableLines()) {
-      final String trimmed = line.trim();
-
-      if (trimmed.startsWith("[{") && trimmed.endsWith("}]")) {
-        LOG.info(trimmed);
+      if (DaemonApi.parseAndValidateDaemonEvent(line.trim()) != null) {
+        if (FlutterSettings.getInstance().isVerboseLogging()) {
+          LOG.info(line.trim().trim());
```
Comment (jacob314) [inline src/io/flutter/run/daemon/DaemonConsoleView.java]: extra call to trim :)

File: src/io/flutter/run/daemon/DaemonEvent.java (hunk 6 above)
Comment (devoncarew) [reply src/io/flutter/run/daemon/DaemonEvent.java]: Logging an error here will actually display a prompt to the user (a crash occurred...). This is just bad data from the wire, not a programmer error (so on the order of not being able to read a file, not an NPE), so I won't want to show a dialog.

File: testSrc/unit/io/flutter/run/daemon/DaemonApiTest.java (hunk 7 above)
Comment (devoncarew) [reply testSrc/unit/io/flutter/run/daemon/DaemonApiTest.java]: 👍 will do

File: src/io/flutter/run/daemon/DaemonConsoleView.java (hunk 8 above)
Comment (devoncarew) [reply src/io/flutter/run/daemon/DaemonConsoleView.java]: I'll trim one out

File: src/io/flutter/run/daemon/DaemonApi.java (hunk 1 above)
Comment (devoncarew) [reply src/io/flutter/run/daemon/DaemonApi.java]: updated the wording

File: src/io/flutter/run/daemon/DaemonApi.java (hunk 2 above)
Comment (devoncarew) [reply src/io/flutter/run/daemon/DaemonApi.java]: ==> `message`

File: src/io/flutter/run/daemon/DaemonApi.java (hunk 3 above)
Comment (devoncarew) [reply src/io/flutter/run/daemon/DaemonApi.java]: good question - in this case we do know that we won't have leading ws

File: src/io/flutter/run/daemon/DaemonConsoleView.java (hunk 4 above)
Comment (devoncarew) [reply src/io/flutter/run/daemon/DaemonConsoleView.java]: updated: `this removes any single newline that occurred`

File: src/io/flutter/run/daemon/DaemonEvent.java (hunk 5 above)
Comment (devoncarew) [reply src/io/flutter/run/daemon/DaemonEvent.java]: 👍, updated

**Comment by devoncarew** [comment]
updated
